File: solutions/typescript/complex-numbers/1/complex-numbers.ts
export class ComplexNumber {
  // Usamos modificadores privados para evitar conflictos con los getters
  private readonly _real: number;
  private readonly _imag: number;

  constructor(real: number, imaginary: number) {
    this._real = real;
    this._imag = imaginary;
  }

  // Getters para acceder a las partes del número
  public get real(): number {
    return this._real;
  }

  public get imag(): number {
    return this._imag;
  }

  // Suma: (a + c) + (b + d)i
  public add(other: ComplexNumber): ComplexNumber {
    return new ComplexNumber(this.real + other.real, this.imag + other.imag);
  }

  // Resta: (a - c) + (b - d)i
  public sub(other: ComplexNumber): ComplexNumber {
    return new ComplexNumber(this.real - other.real, this.imag - other.imag);
  }

  // Multiplicación: (a*c - b*d) + (b*c + a*d)i
  public mul(other: ComplexNumber): ComplexNumber {
    const realPart = this.real * other.real - this.imag * other.imag;
    const imagPart = this.imag * other.real + this.real * other.imag;
    return new ComplexNumber(realPart, imagPart);
  }

  // División: requiere usar el cuadrado del denominador (c^2 + d^2)
  public div(other: ComplexNumber): ComplexNumber {
    const denominator = other.real * other.real + other.imag * other.imag;
    const realPart = (this.real * other.real + this.imag * other.imag) / denominator;
    const imagPart = (this.imag * other.real - this.real * other.imag) / denominator;
    return new ComplexNumber(realPart, imagPart);
  }

  // Valor absoluto: sqrt(a^2 + b^2)
  public get abs(): number {
    return Math.sqrt(this.real * this.real + this.imag * this.imag);
  }

  // Conjugado: a - bi
  public get conj(): ComplexNumber {
    return new ComplexNumber(this.real, this.imag === 0 ? 0 : -this.imag);
  }

  // Exponencial: e^a * (cos(b) + i * sin(b))
  public get exp(): ComplexNumber {
    const expReal = Math.exp(this.real);
    return new ComplexNumber(
      expReal * Math.cos(this.imag),
      expReal * Math.sin(this.imag)
    );
  }
}

File: solutions/ruby/complex-numbers/1/complex_numbers.rb
class ComplexNumber
  attr_reader :real, :imaginary

  def initialize(real, imaginary = 0)
    @real = real.to_f
    @imaginary = imaginary.to_f
  end

  # Suma: (a + c) + (b + d)i
  def +(other)
    ComplexNumber.new(real + other.real, imaginary + other.imaginary)
  end

  # Resta: (a - c) + (b - d)i
  def -(other)
    ComplexNumber.new(real - other.real, imaginary - other.imaginary)
  end

  # Multiplicación: (a*c - b*d) + (b*c + a*d)i
  def *(other)
    new_real = real * other.real - imaginary * other.imaginary
    new_imag = imaginary * other.real + real * other.imaginary
    ComplexNumber.new(new_real, new_imag)
  end

  # División: ((a*c + b*d) / denominator) + ((b*c - a*d) / denominator)i
  def /(other)
    denominator = other.real**2 + other.imaginary**2
    new_real = (real * other.real + imaginary * other.imaginary) / denominator
    new_imag = (imaginary * other.real - real * other.imaginary) / denominator
    ComplexNumber.new(new_real, new_imag)
  end

  # Conjugado: a - bi
  def conjugate
    ComplexNumber.new(real, -imaginary)
  end

  # Valor Absoluto (Módulo): sqrt(a^2 + b^2)
  def abs
    Math.sqrt(real**2 + imaginary**2)
  end

  # Exponenciación de e: e^a * cos(b) + (e^a * sin(b))i
  def exp
    factor = Math.exp(real)
    ComplexNumber.new(factor * Math.cos(imaginary), factor * Math.sin(imaginary))
  end

  # Método de comparación para que los asserts de los tests funcionen correctamente
  def ==(other)
    return false unless other.is_a?(ComplexNumber)
    
    # Usamos delta/tolerancia por imprecisiones decimales de punto flotante (Float)
    (real - other.real).abs < 1e-9 && (imaginary - other.imaginary).abs < 1e-9
  end
end